File: app/src/main/java/mishka/rssreader/ui/feed/FeedActivity.java
package mishka.rssreader.ui.feed;

import android.arch.lifecycle.Observer;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;

import java.util.Iterator;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.OrderedCollectionChangeSet;
import io.realm.OrderedRealmCollectionChangeListener;
import io.realm.Realm;
import io.realm.RealmChangeListener;
import io.realm.RealmResults;
import mishka.rssreader.R;
import mishka.rssreader.data.model.RealmRssItem;
import mishka.rssreader.di.component.DaggerFeedComponent;
import mishka.rssreader.di.component.FeedComponent;
import mishka.rssreader.ui.BaseActivity;

public class FeedActivity extends BaseActivity {

    @BindView(R.id.feed_recycler_view)
    RecyclerView feedRecyclerView;

    @BindView(R.id.refresh)
    SwipeRefreshLayout refreshLayout;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Inject
    FeedAdapter adapter;

    @Inject
    FeedViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);

        setSupportActionBar(toolbar);

        FeedComponent feedComponent = DaggerFeedComponent.builder().applicationComponent(getApplicationComponent()).build();

        feedComponent.inject(this);

        feedRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        feedRecyclerView.setAdapter(adapter);

        adapter.setPosts(viewModel.getRealmItems());

        viewModel.getRealmItems().addChangeListener(new RealmChangeListener<RealmResults<RealmRssItem>>() {
            @Override
            public void onChange(RealmResults<RealmRssItem> realmRssItems) {
                adapter.setPosts(realmRssItems);
            }
        });

        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                viewModel.update();
                refreshLayout.setRefreshing(false);
            }
        });
    }
}

File: app/src/main/java/mishka/rssreader/ui/feed/FeedActivity.kt
package mishka.rssreader.ui.feed;

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.toolbar.*
import mishka.rssreader.R


import javax.inject.Inject

import mishka.rssreader.data.model.RssItem
import mishka.rssreader.di.component.DaggerFeedComponent
import mishka.rssreader.di.component.FeedComponent
import mishka.rssreader.ui.BaseActivity

class FeedActivity : BaseActivity() {

    @Inject
    lateinit var adapter: FeedAdapter

    @Inject
    lateinit var viewModel: FeedViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)

        val component = DaggerFeedComponent.builder().applicationComponent(getApplicationComponent()).build()

        component.inject(this)

        rvFeed.layoutManager = LinearLayoutManager(this)

        rvFeed.adapter = adapter

        viewModel.posts.observe(this, object : Observer<List<RssItem>> {
            override fun onChanged(posts: List<RssItem>?) {
                adapter.posts = posts!!
            }
        })

        refresh.setOnRefreshListener(object : SwipeRefreshLayout.OnRefreshListener {
            override fun onRefresh() {
                viewModel.update()
                refresh.isRefreshing = false
            }
        })

        hideSettings()
    }

    fun notifyCurrentChannelChanged() {
        viewModel.update()
        hideSettings()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.feed_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    var isSettingsOpened = false

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.settings -> if (isSettingsOpened) hideSettings() else showSettings()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showSettings() {
        channelSettings.view?.visibility = View.VISIBLE
        isSettingsOpened = true
    }

    private fun hideSettings() {
        channelSettings.view?.visibility = View.GONE
        isSettingsOpened = false
    }

}
